### ISSUE TITLE
Password protect "Stay logged in" feature with pin or simpler password

### ISSUE THREAD

**Issue by opensource21** [turtl/tracker#121]
It would be nice to have a simple login on securce machines. For the internet I prefer to have a strong password. How ever on devices like my desktop PC I like it to have quick access. So it would be nice to save the master key (like in #89) and save it with a simpler password. If the password is wrong then delete the safe data.

**Comment by orthecreedence** [comment]
Agree, this would be a cool feature if your desktop has hard drive encryption, and could be a nice middle ground between having to type the full password and full relying on the "Stay logged in" feature's use of the underlying system's security.

**Comment by thecount2a** [comment]
I would really like this feature to improve the security of this app! I love Turtl so far, I tried it about 1 year ago and it was somewhat unstable but I tried it again recently and version 0.7 really seems like a brand new experience! Thank you for all of your hard work on this!

I would like to use a really long password (basically an encryption key) but then have a second password that would be requested every time I open the app. This would mitigate the attack where an adversary is able to compromise my shorter, memorized password but then has no chance of guessing the longer one. Of course I could just set my password to a very long password but then anybody who walked up to my computer while the screen was unlocked could launch the turtl program and have access to my note database with whatever passwords it might contain. Of course it comes with the trade-off of needing this very long password to provision new devices but I am easily willing to accept this.

One way to implement this would be to offer an "account key" field within the Advanced Settings. Then, whenever turtl is generating the encryption key to open the database, the user's password is appended to this "account key". This happens during account creation and during login. I'm assuming that the Advanced Settings are persisted to disk. Then the users who choose to fully trust the disk on their device can use the "Stay logged in" feature to save both the password AND the account key and users who don't fully trust their system (say on a non-mobile platform), would choose not to use "Stay logged in" and their devices would still persist the account key in the Advanced Settings page.  Any users who forget either their password or their account key would lose their data but since they chose to use the Advanced Settings option to set an account key in the first place, one would hope they would back up this key somewhere.

**Comment by orthecreedence** [comment]
@thecount2a Great ideas! Thanks for the feedback (and also, I'm glad the new version of the app is working better for you, that's great to hear).

I like the idea of splitting out the pin/account key from the main password.

**Comment by opensource21** [comment]
What do you think if we protect the key at android devices with a fingerprint? If it should be optional the JS-Part must add this input, because it's hard to make gui and storage native in a cordova app.

**Comment by orthecreedence** [comment]
I love that idea. We've gotten requests for full-fingerprint login, however I don't think it would be possible to accurately generate the master key from only a fingerprint. However, using a fingerprint to protect the "stay logged in" pin instead of generate the master key could work.

- https://www.npmjs.com/package/cordova-plugin-android-fingerprint-auth
- https://www.npmjs.com/package/cordova-plugin-fingerprint-id
